### ISSUE TITLE
In view, label of date and number fields isn't read

### ISSUE THREAD

**Issue by avernet** [orbeon/orbeon-forms#6707]
See how how, for the first control, the `<span>1/14/2025</span>` is missing a number of the `aria-` attributes, while those are present to `<span>52</span>` for the second control.

![Screenshot 2025-01-06 at 3 53 35 PM](https://github.com/user-attachments/assets/a908fd7e-6e20-4fb3-8fc6-5cabb80db810)

[Form in Form Builder on demo](https://demo.orbeon.com/demo/fr/orbeon/builder/edit/afe6151cf8a85b3891bb739e1513f6ef2671a10a)
[+1 from customer](https://3.basecamp.com/3600924/buckets/1956846/messages/7175826989)

**Comment by avernet** [comment]
I updated the title and example to show that this problem also occurs with the number field. Hopefully, the solution for both the date and number fields will be the same.

**Comment by ebruchez** [comment]
We store a single `LHHAAnalysis.lhhaPlacementType`, which is then used in a few places. This is used:

- `Local.directTargetControl`
- `External.forRepeatNestingOpt`
- `isForRepeat`
- `isLocal`

However, if we restrict what `xxbl:label-for` can do, this might work:

- single target: like now
- multiple targets
    - cannot be `Local`
    - `External` must not point to a nested repeat (all must be non-repeated)
    - but they can mix and match `LhhaControlRef.Control` and `LhhaControlRef.PrefixedId` (as `fr:date` would)

Thinking about this, I am not sure it will work given the number of cases to handle! And annoyingly, we would be implementing this just to handle two exclusive branches: regular, and static-readonly. It does not, for now at lesat, satisfy other valid use cases, and that would come at the price of complexity.

Can we explore the alternative of updating components with a static-readonly annotation? What it would take:

- in `components.xsl`/`controls.xsl`: add an attribute to controls within grids
    - for example  `fr:static-readonly="true"`
    - this is ok, because `new`/`edit` and `view`/`pdf` are separate compiled forms
- create new XBL bindings that bind to:
    - `fr|date[fr:static-readonly = true], xf|input[fr:static-readonly = true]:xxf-type('xs:date')
    - same for `fr:number`, maybe others
    - Q: Do we support the binding logic?
        - #4370 seems to say that we support the representation in `BindingDescriptor`
        - but we don't support this in Form Builder (which is ok for this current issue)
        - but do we support the binding logic at all? we imply that not; if not, would that be easy?

**Comment by ebruchez** [comment]
`BindingIndex.findMostSpecificBinding()`. `BindingDescriptor`, right now, only holds a *single* optional attribute selector. This would work for a plain `fr:date` and `fr:number`, since they don't match on anything else, but it wouldn't work for `fr:dropdown-date`, in particular, which has `xf|input:xxf-type('xs:date')[appearance ~= dropdowns]`.

In [this Gist](https://gist.github.com/ebruchez/4cb21b8f00bded567948f8c2d1453344) I show which bindings based on attributes we have right now for the Controls form. We bind on:

- `appearance`
- `mediatype` (`xf:textarea`)
- `repeat`, `bind`, `nodeset`, `ref` (`fr:grid`)

**Comment by ebruchez** [comment]
So we do have the ability to match on different attributes, just a single attribute at a time. The code does check all attributes of the element for a match.

**Comment by ebruchez** [comment]
Issue: right now we determine `xxf:readonly-appearance="static"` using an XForms AVT, which is lazily evaluated. So we don't know it yet at form transformation time. Yet we should be able to know it, right?

```
xxf:readonly-appearance="{{
    for $mode in fr:mode()
    return
        if (
            $mode = 'view' or (
                $mode = ('pdf', 'test-pdf', 'email') and not(fr:use-pdf-template())
            )
        ) then
            'static'
        else
            'dynamic'
}}"
```

The mode as known by XSLT and the mode as known by XForms can differ:

```
<xf:bind
    ref="instance('fr-parameters-instance')/mode"
    xxf:default="
        if (. = ('tiff', 'test-pdf')) then
            'pdf'
        else if (. = ('export', 'excel-export') and xxf:is-blank(../document)) then
            'new'
        else if (. = ('export', 'excel-export') and xxf:non-blank(../document)) then
            'edit'
        else
            ."/>
```

(Based on this, testing on `'test-pdf'` in the first case above should not be needed.)

`fr:use-pdf-template()` depends on:

- content of `fr-form-attachments` (so form definition, which is ok)
- `req.getMethod`
- `req.getFirstParamAsString` for `fr-use-pdf-template`

The latter two are annoying, because that means we depend on parameters and method. Does that mean we'd possibly generate two different compiled forms depending on whether we are doing a `POST` or not? This is unclear.

**Comment by ebruchez** [comment]
After further discussion. I dislike the first approach of supporting multiple references for `label-for`, because that's not a feature we want in itself. In practice, we always only want a single target for `label-for`.

The annotation approach might be better, but it is work. In the longer term, we don't want the PDF template to work the way it does (#1573). But in the meanwhile, it is probably ok to modify `components.xsl` to depend on `fr-use-pdf-template` and on the HTTP method. This would cause a different compiled version for the form. The only issue is it is not an insignificant change.

- [ ] move `static-readonly` detection to `components.xsl`
- [ ] add annotations on controls within grids
- [ ] create separate XBL components for static-readonly appearances of Date, Number, etc.

**Comment by ebruchez** [comment]
One issue with `fr:date`: if `fr:component-param-value('native-picker') = 'always'`, in static readonly, we still output the HTML `xh:input` "to have consistent output format". Can figure this out outside of `fr:date`? Not ideal, but maybe the only way?

**Comment by ebruchez** [comment]
- [ ] check why we have an `aria-labelledby` pointing to a `<span>`, instead of `<label>` pointing to `<input>`

**Comment by ebruchez** [comment]
In `handleLabel()`, we have the following to output a `<span>` in static readonly mode instead of a `<label>`:

```scala
protected def handleLabel(lhhaAnalysis: LHHAAnalysis): Unit =
  handleLabelHintHelpAlert(
    lhhaAnalysis            = lhhaAnalysis,
    controlEffectiveIdOpt   = XFormsBaseHandler.isStaticReadonly(currentControl) option getEffectiveId,
    forEffectiveIdWithNsOpt = getForEffectiveIdWithNs(lhhaAnalysis),
    requestedElementNameOpt = XFormsBaseHandler.isStaticReadonly(currentControl) option "span",
    control                 = currentControl
  )
```

On the other side, we use `handleAriaByAtts()` to place `ariea-labelledby` and others, but it's unclear what the logic is here.

**Comment by ebruchez** [comment]
Discussion with @avernet: with the current solution, the decision about whether to show a static readonly date or time, due to the native picker logic, is taken externally from `fr:date` and `fr:time`, which is not ideal. We could possibly improve this in the future by providing more attributes on the controls: instead or in addition to `static-readonly`, we could maybe add:

- `mode`
- `use-pdf-template`
- `native-picker` (resolved using form/properties as well)

and have the binding selector use the combination of these extra attributes to make the decision.

**Comment by ebruchez** [comment]
Included in 2024.1.1. However, a backport to 2023.1.x might be too difficult.

**Comment by avernet** [comment]
> Included in 2024.1.1. However, a backport to 2023.1.x might be too difficult.

I removed this from the 2023.1.8 project.
